[PyTorch] Fix const correctness for resize native functions

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #55691 [PyTorch] Migrate add operators to borrow in TensorIteratorBase
* #55690 [PyTorch] Support borrowing in/out Tensors in TensorIterator
* #55940 [PyTorch] Always update op.current_dtype in TensorIteratorBase::set_output
* #55420 [PyTorch] Remove non-const TensorIterator::tensor() method
* #55258 [PyTorch] Format generated structured kernels code better
* **#55351 [PyTorch] Fix const correctness for resize native functions**

We incorrectly used `Tensor&` to mean "the underlying
TensorImpl cannot be changed", as explained in
https://github.com/zdevito/ATen/issues/27#issuecomment-330717839 .
This diff gets us on the path to fixing this problem: we have an
incremental way to fix individual native functions so that we can
apply any handwritten fixes a few at a time. It gets the migration
started with the `resize` family of native functions.

Differential Revision: [D27583983](https://our.internmc.facebook.com/intern/diff/D27583983/)

swolchok committed:
[PyTorch] Fix const correctness for resize native functions

We incorrectly used `Tensor&` to mean "the underlying
TensorImpl cannot be changed", as explained in
https://github.com/zdevito/ATen/issues/27#issuecomment-330717839 .
This diff gets us on the path to fixing this problem: we have an
incremental way to fix individual native functions so that we can
apply any handwritten fixes a few at a time. It gets the migration
started with the `resize` family of native functions.

Differential Revision: [D27583983](https://our.internmc.facebook.com/intern/diff/D27583983/)

[ghstack-poisoned]

swolchok commented on tools/autograd/gen_variable_type.py:
```diff
@@ -688,7 +689,8 @@ def emit_check_inplace() -> List[str]:
     body.append(declare_returned_variables(f))
 
-    body.append(emit_call(f, unpacked_bindings))
+    with native_function_manager(f):
```
1) It is documented as being mandatory in order to call API functions on line 29 of context.py, we just never enforced that mandate
2) It registers the `local()` state by calling `local.parametrize` with `f.use_const_ref_for_mutable_tensors`; see changes to local.py and context.py

ezyang replied:
It seems like it would be better to apply the manager to the entirety of `emit_body`, so I'm curious why your decorator doesn't apply here

swolchok replied:
nice catch, it should

swolchok committed:
forgot to add tools/autograd/content.py on "[PyTorch] Fix const correctness for resize native functions"


We incorrectly used `Tensor&` to mean "the underlying
TensorImpl cannot be changed", as explained in
https://github.com/zdevito/ATen/issues/27#issuecomment-330717839 .
This diff gets us on the path to fixing this problem: we have an
incremental way to fix individual native functions so that we can
apply any handwritten fixes a few at a time. It gets the migration
started with the `resize` family of native functions.

Differential Revision: [D27583983](https://our.internmc.facebook.com/intern/diff/D27583983/)

[ghstack-poisoned]

swolchok committed:
fix runtime crashes due to manual autograd/tracer code on "[PyTorch] Fix const correctness for resize native functions"


We incorrectly used `Tensor&` to mean "the underlying
TensorImpl cannot be changed", as explained in
https://github.com/zdevito/ATen/issues/27#issuecomment-330717839 .
This diff gets us on the path to fixing this problem: we have an
incremental way to fix individual native functions so that we can
apply any handwritten fixes a few at a time. It gets the migration
started with the `resize` family of native functions.

Differential Revision: [D27583983](https://our.internmc.facebook.com/intern/diff/D27583983/)

[ghstack-poisoned]

swolchok committed:
fix resize_cuda_ on "[PyTorch] Fix const correctness for resize native functions"


We incorrectly used `Tensor&` to mean "the underlying
TensorImpl cannot be changed", as explained in
https://github.com/zdevito/ATen/issues/27#issuecomment-330717839 .
This diff gets us on the path to fixing this problem: we have an
incremental way to fix individual native functions so that we can
apply any handwritten fixes a few at a time. It gets the migration
started with the `resize` family of native functions.

Differential Revision: [D27583983](https://our.internmc.facebook.com/intern/diff/D27583983/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Fix const correctness for resize native functions"


We incorrectly used `Tensor&` to mean "the underlying
TensorImpl cannot be changed", as explained in
https://github.com/zdevito/ATen/issues/27#issuecomment-330717839 .
This diff gets us on the path to fixing this problem: we have an
incremental way to fix individual native functions so that we can
apply any handwritten fixes a few at a time. It gets the migration
started with the `resize` family of native functions.

Differential Revision: [D27583983](https://our.internmc.facebook.com/intern/diff/D27583983/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Fix const correctness for resize native functions"


We incorrectly used `Tensor&` to mean "the underlying
TensorImpl cannot be changed", as explained in
https://github.com/zdevito/ATen/issues/27#issuecomment-330717839 .
This diff gets us on the path to fixing this problem: we have an
incremental way to fix individual native functions so that we can
apply any handwritten fixes a few at a time. It gets the migration
started with the `resize` family of native functions.

Differential Revision: [D27583983](https://our.internmc.facebook.com/intern/diff/D27583983/)

[ghstack-poisoned]

ezyang commented on tools/codegen/model.py:
```diff
@@ -263,6 +267,11 @@ def from_yaml(ei: Dict[str, object], loc: 'Location') -> 'NativeFunction':
+        use_const_ref_for_mutable_tensors = e.pop('use_const_ref_for_mutable_tensors', False)
+        assert isinstance(use_const_ref_for_mutable_tensors, bool)
+        if "resize" in str(func.name):
+            assert use_const_ref_for_mutable_tensors, "oh no " + func
```
lul

ezyang commented on tools/codegen/model.py:
```diff
@@ -736,21 +746,25 @@ class Annotation:
     # we can conveniently assume it is canonically ordered
     alias_set: Tuple[str, ...]
     is_write: bool
+    is_const_ref_write: bool
```
This guy is dead now, remove it and all the other parsey printey stuff for it

ezyang reviewed (approved):
that's one small step for mankind

swolchok committed:
Update on "[PyTorch] Fix const correctness for resize native functions"


We incorrectly used `Tensor&` to mean "the underlying
TensorImpl cannot be changed", as explained in
https://github.com/zdevito/ATen/issues/27#issuecomment-330717839 .
This diff gets us on the path to fixing this problem: we have an
incremental way to fix individual native functions so that we can
apply any handwritten fixes a few at a time. It gets the migration
started with the `resize` family of native functions.

Differential Revision: [D27583983](https://our.internmc.facebook.com/intern/diff/D27583983/)

[ghstack-poisoned]

swolchok committed:
remove dead code. add decorator. on "[PyTorch] Fix const correctness for resize native functions"


We incorrectly used `Tensor&` to mean "the underlying
TensorImpl cannot be changed", as explained in
https://github.com/zdevito/ATen/issues/27#issuecomment-330717839 .
This diff gets us on the path to fixing this problem: we have an
incremental way to fix individual native functions so that we can
apply any handwritten fixes a few at a time. It gets the migration
started with the `resize` family of native functions.

Differential Revision: [D27583983](https://our.internmc.facebook.com/intern/diff/D27583983/)

[ghstack-poisoned]

swolchok committed:
rebase on "[PyTorch] Fix const correctness for resize native functions"


We incorrectly used `Tensor&` to mean "the underlying
TensorImpl cannot be changed", as explained in
https://github.com/zdevito/ATen/issues/27#issuecomment-330717839 .
This diff gets us on the path to fixing this problem: we have an
incremental way to fix individual native functions so that we can
apply any handwritten fixes a few at a time. It gets the migration
started with the `resize` family of native functions.

Differential Revision: [D27583983](https://our.internmc.facebook.com/intern/diff/D27583983/)

[ghstack-poisoned]

swolchok committed:
fix newly-added resize_as_sparse_ on "[PyTorch] Fix const correctness for resize native functions"


We incorrectly used `Tensor&` to mean "the underlying
TensorImpl cannot be changed", as explained in
https://github.com/zdevito/ATen/issues/27#issuecomment-330717839 .
This diff gets us on the path to fixing this problem: we have an
incremental way to fix individual native functions so that we can
apply any handwritten fixes a few at a time. It gets the migration
started with the `resize` family of native functions.

Differential Revision: [D27583983](https://our.internmc.facebook.com/intern/diff/D27583983/)

[ghstack-poisoned]

swolchok committed:
rebase w/conflicts on "[PyTorch] Fix const correctness for resize native functions"


We incorrectly used `Tensor&` to mean "the underlying
TensorImpl cannot be changed", as explained in
https://github.com/zdevito/ATen/issues/27#issuecomment-330717839 .
This diff gets us on the path to fixing this problem: we have an
incremental way to fix individual native functions so that we can
apply any handwritten fixes a few at a time. It gets the migration
started with the `resize` family of native functions.

Differential Revision: [D27583983](https://our.internmc.facebook.com/intern/diff/D27583983/)

[ghstack-poisoned]

swolchok committed:
fix resize_as_sparse_csr_ on "[PyTorch] Fix const correctness for resize native functions"


We incorrectly used `Tensor&` to mean "the underlying
TensorImpl cannot be changed", as explained in
https://github.com/zdevito/ATen/issues/27#issuecomment-330717839 .
This diff gets us on the path to fixing this problem: we have an
incremental way to fix individual native functions so that we can
apply any handwritten fixes a few at a time. It gets the migration
started with the `resize` family of native functions.

Differential Revision: [D27583983](https://our.internmc.facebook.com/intern/diff/D27583983/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Fix const correctness for resize native functions"


We incorrectly used `Tensor&` to mean "the underlying
TensorImpl cannot be changed", as explained in
https://github.com/zdevito/ATen/issues/27#issuecomment-330717839 .
This diff gets us on the path to fixing this problem: we have an
incremental way to fix individual native functions so that we can
apply any handwritten fixes a few at a time. It gets the migration
started with the `resize` family of native functions.

Differential Revision: [D27583983](https://our.internmc.facebook.com/intern/diff/D27583983/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Fix const correctness for resize native functions"


We incorrectly used `Tensor&` to mean "the underlying
TensorImpl cannot be changed", as explained in
https://github.com/zdevito/ATen/issues/27#issuecomment-330717839 .
This diff gets us on the path to fixing this problem: we have an
incremental way to fix individual native functions so that we can
apply any handwritten fixes a few at a time. It gets the migration
started with the `resize` family of native functions.

Differential Revision: [D27583983](https://our.internmc.facebook.com/intern/diff/D27583983/)

[ghstack-poisoned]

ailzhang reviewed (approved):
Thanks for waiting! @JackCaoG has prepared a patch for this PR(thanks!) and it's good to go!

swolchok committed:
Update on "[PyTorch] Fix const correctness for resize native functions"


We incorrectly used `Tensor&` to mean "the underlying
TensorImpl cannot be changed", as explained in
https://github.com/zdevito/ATen/issues/27#issuecomment-330717839 .
This diff gets us on the path to fixing this problem: we have an
incremental way to fix individual native functions so that we can
apply any handwritten fixes a few at a time. It gets the migration
started with the `resize` family of native functions.

Differential Revision: [D27583983](https://our.internmc.facebook.com/intern/diff/D27583983/)

[ghstack-poisoned]

swolchok committed:
fix build at bottom of stack on "[PyTorch] Fix const correctness for resize native functions"


We incorrectly used `Tensor&` to mean "the underlying
TensorImpl cannot be changed", as explained in
https://github.com/zdevito/ATen/issues/27#issuecomment-330717839 .
This diff gets us on the path to fixing this problem: we have an
incremental way to fix individual native functions so that we can
apply any handwritten fixes a few at a time. It gets the migration
started with the `resize` family of native functions.

Differential Revision: [D27583983](https://our.internmc.facebook.com/intern/diff/D27583983/)

[ghstack-poisoned]

swolchok committed:
fix python formatting on "[PyTorch] Fix const correctness for resize native functions"


We incorrectly used `Tensor&` to mean "the underlying
TensorImpl cannot be changed", as explained in
https://github.com/zdevito/ATen/issues/27#issuecomment-330717839 .
This diff gets us on the path to fixing this problem: we have an
incremental way to fix individual native functions so that we can
apply any handwritten fixes a few at a time. It gets the migration
started with the `resize` family of native functions.

Differential Revision: [D27583983](https://our.internmc.facebook.com/intern/diff/D27583983/)

[ghstack-poisoned]

swolchok commented:
was not able to repro CUDAPytorchToCaffe2.MutualResizes failure locally on a machine with a GPU, and the failing line is before it executes any resize operations, so I'm going to go ahead and call this test result landable.

swolchok committed:
Update on "[PyTorch] Fix const correctness for resize native functions"


We incorrectly used `Tensor&` to mean "the underlying
TensorImpl cannot be changed", as explained in
https://github.com/zdevito/ATen/issues/27#issuecomment-330717839 .
This diff gets us on the path to fixing this problem: we have an
incremental way to fix individual native functions so that we can
apply any handwritten fixes a few at a time. It gets the migration
started with the `resize` family of native functions.

Differential Revision: [D27583983](https://our.internmc.facebook.com/intern/diff/D27583983/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Fix const correctness for resize native functions"


We incorrectly used `Tensor&` to mean "the underlying
TensorImpl cannot be changed", as explained in
https://github.com/zdevito/ATen/issues/27#issuecomment-330717839 .
This diff gets us on the path to fixing this problem: we have an
incremental way to fix individual native functions so that we can
apply any handwritten fixes a few at a time. It gets the migration
started with the `resize` family of native functions.

Differential Revision: [D27583983](https://our.internmc.facebook.com/intern/diff/D27583983/)

[ghstack-poisoned]
